### PR TITLE
9948 stop orders linked to order or position 2

### LINKED ISSUE #9948: Position/order linked stop orders
This issue is to implement the changes to stop orders and laid out in the below spec changes:

- Attach stop to order (volume “activates” as order trades, unactivated is cancelled if order no longer on book or parked)
- Attach stop to position (volume grows/shrinks with position, cancelled if position 0 or changes direction)

Specs:

- https://github.com/vegaprotocol/specs/issues/2025

### PR DESCRIPTION
**Pull request by jeremyletang**
rebased https://github.com/vegaprotocol/vega/pull/10028

close #9948

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -3130,6 +3130,16 @@ enum StopOrderRejectionReason {
   REJECTION_REASON_STOP_ORDER_NOT_CLOSING_THE_POSITION
 }
 
+"Stop order size override settings"
+enum StopOrderSizeOverrideSetting {
+  "The size override has not been specified, this should never happen!"
+  SIZE_OVERRIDE_SETTING_UNSPECIFIED
+  "No size override"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:3137]:
```suggestion
  "No size override, the size within the contained normal order submission will be used
```

File: protos/sources/vega/vega.proto [hunk 2]
```diff
@@ -7,6 +7,20 @@ import "vega/markets.proto";
 option go_package = "code.vegaprotocol.io/vega/protos/vega";
 
 message StopOrder {
+  message SizeOverrideValue {
+    // Scaling percentage of the current positions size
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:11]:
```suggestion
    // Scaling percentage of the current position’s size
```

File: core/execution/stoporders/stop_orders.go [hunk 3]
```diff
@@ -107,6 +108,42 @@ func (p *Pool) Settled() []*types.StopOrder {
...
+func (p *Pool) CheckDirection(positions *positions.SnapshotEngine) []*types.StopOrder {
+	toCancel := []*types.StopOrder{}
+	for partyID, orders := range p.orders {
+		pos, ok := positions.GetPositionByPartyID(partyID)
+		if !ok {
+			continue
+		}
+		for _, order := range orders {
+			if order.SizeOverrideSetting == types.StopOrderSizeOverrideSettingPosition {
+				if pos.Size() == 0 {
```
Comment (EVODelavega) [inline core/execution/stoporders/stop_orders.go:120]: too many `else`'s (no need after continue, the last `else` with just an `if` also is a bit messy

File: protos/sources/vega/commands/v1/commands.proto [hunk 4]
```diff
@@ -48,6 +48,10 @@ message StopOrderSetup {
   optional int64 expires_at = 2;
   // Strategy to adopt if the expiry time is reached.
   optional vega.StopOrder.ExpiryStrategy expiry_strategy = 3;
+  // Is this order linked to something to derive the order size
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:51]:
```suggestion
  // Indicates if this order is linked to an order or position to derive the order size
```

File: protos/sources/vega/vega.proto [hunk 5]
```diff
@@ -7,6 +7,20 @@ import "vega/markets.proto";
...
+    string percentage = 1;
+  }
+
+  enum SizeOverrideSetting {
+    // Never valid
+    SIZE_OVERRIDE_SETTING_UNSPECIFIED = 0;
+    // No override
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:18]:
```suggestion
    // No override, the size within the contained normal order submission will be used
```

**Review by EVODelavega — APPROVED** [review]

**Review by candida-d — APPROVED** [review]

**Pull request merged**